[quant] Add optimized approach to calculate qparams for qembedding_bag

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45149 [quant] Add optimized approach to calculate qparams for qembedding_bag**

Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23848060](https://our.internmc.facebook.com/intern/diff/D23848060)

supriyar committed:
[quant] Add optimized approach to calculate qparams for qembedding_bag
Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4256,6 +4256,10 @@
   use_c10_dispatcher: full
   variants: function
 
+- func: choose_qparams_optimized(Tensor input, int n, int n_bins, float ratio, float xmin, float xmax, int bit_rate) -> (float, float)
```
NIT: Can we call this bit_width instead of bit_rate?
Also, do we distinguish between fp16 and fp32 outputs?

supriyar replied:
This function returns the optimized xmin and xmax values currently. The calculation of scale and zero point happens at the callsite,.

raghuramank100 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -232,5 +232,93 @@ std::tuple<double, int64_t> _choose_qparams_per_tensor(
+    double xmax,
+    int64_t bit_rate) {
+
+  xmin = (float) xmin;
```
Do we need xmin and xmax as inputs? Can we compute them from the input tensor to start the optimization?

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -94,6 +94,7 @@ namespace {
 // Note - This is a temporary pack function for embedding bag which quantizes
 // and packs the float weight tensor. In the next step it will be replaced by a
 // quantize and pack function once we support FP scale and FP zero_point
+template <bool OPTIMIZED_QPARAMS = false>
```
Does this need to be templatized? Can we have this as an argument instead?

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23848060](https://our.internmc.facebook.com/intern/diff/D23848060)

[ghstack-poisoned]

raghuramank100 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -404,6 +404,7 @@ def get_cpp_formal(arg, ensure_temp_safe=True):
     'std::tuple<Tensor,Tensor,Tensor,Tensor,int64_t>',
     'std::tuple<Tensor,Tensor,double,Tensor,int64_t>',
     'std::tuple<double,int64_t>',
+    'std::tuple<double,double>',
```
Is this related to your change?

supriyar replied:
yes, this adds a new combination of return types that is not currently supported.

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2722,7 +2721,7 @@ def _test_embedding_bag_unpack_fn(self, pack_fn, unpack_fn, num_embeddings, embe
             num_embeddings, embedding_dim)) + 1).astype(np.float32))
 
-        w_packed = pack_fn(weights)
+        w_packed = pack_fn(weights, optimized_qparams=True)
```
Should we test for both cases of optimized_qparams for completeness?

supriyar replied:
Yes we should, not sure how I missed this.

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qembeddingbag_prepack.cpp:
```diff
@@ -94,7 +94,7 @@ namespace {
 // and packs the float weight tensor. In the next step it will be replaced by a
 // quantize and pack function once we support FP scale and FP zero_point
-Tensor qembeddingbag_byte_prepack(const Tensor& weight) {
+Tensor qembeddingbag_byte_prepack(const Tensor& weight, bool optimized_qparams) {
```
This function is not doing optimized_qparam calculation, can you please add that?

supriyar replied:
In caffe2 we currently don't have optimized_qparam calculation for `byte_prepack`. Do we need it in PT?

raghuramank100 commented on aten/src/ATen/native/quantized/library.cpp:
```diff
@@ -105,11 +105,11 @@ TORCH_LIBRARY(quantized, m) {
   m.def("embedding_bag_prepack(Tensor weight) -> __torch__.torch.classes.quantized.EmbeddingPackedParamsBase W_prepack");
   m.def("embedding_bag_unpack(__torch__.torch.classes.quantized.EmbeddingPackedParamsBase W_prepack) -> Tensor W_origin");
-  m.def("embedding_bag_byte_prepack(Tensor weight) -> Tensor");
+  m.def("embedding_bag_byte_prepack(Tensor weight, bool optimized_qparams=False) -> Tensor");
```
When we enable this at the python level, how does the user control which observer is used for embeddings? i.e should we change the default for this op to have optimized_qparams=True?

supriyar replied:
We can have a separate observer that calls the aten op to calculate optimized qparams. 
I'm not sure if setting this to true will have any implications on PyPer perf.

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23848060](https://our.internmc.facebook.com/intern/diff/D23848060)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23848060](https://our.internmc.facebook.com/intern/diff/D23848060)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23848060](https://our.internmc.facebook.com/intern/diff/D23848060)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23848060](https://our.internmc.facebook.com/intern/diff/D23848060)

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add optimized approach to calculate qparams for qembedding_bag"


Summary:
The choose_qparams_optimized calculates the the optimized qparams.
It uses a greedy approach to nudge the min and max and calculate the l2 norm
  and tries to minimize the quant error by doing `torch.norm(x-fake_quant(x,s,z))`
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23848060](https://our.internmc.facebook.com/intern/diff/D23848060)

[ghstack-poisoned]
